internal/clients/content_sources/client.go

```diff
@@ -48,26 +50,79 @@ func (csc *ContentSourcesClient) request(method, url string, headers map[string]
 	return csc.client.Do(req)
 }
 
-// returns ApiRepositoryCollectionResponse
-func (csc *ContentSourcesClient) GetRepositories(ctx context.Context, repoURLs []string, external bool) (*http.Response, error) {
+func (csc *ContentSourcesClient) fetchRepositories(ctx context.Context, repoURLs []string, repoIDs []string, external bool) (*ApiRepositoryCollectionResponse, error) {
 	id, ok := identity.GetIdentityHeader(ctx)
 	if !ok {
 		return nil, fmt.Errorf("Unable to get identity from context")
 	}
 
 	csReposURL := csc.url.JoinPath("repositories/")
 	queryValues := csReposURL.Query()
-	queryValues.Add("url", strings.Join(repoURLs, ","))
+	if len(repoURLs) > 0 {
+		queryValues.Add("url", strings.Join(repoURLs, ","))
+	} else if len(repoIDs) > 0 {
+		queryValues.Add("uuid", strings.Join(repoIDs, ","))
+	} else {
+		return nil, fmt.Errorf("At least one repo url or repo id needs to be given")
+	}
 	if external {
-		queryValues.Add("origin", "external")
+		queryValues.Add("origin", "external,upload")
 	} else {
 		queryValues.Add("origin", "red_hat")
 	}
 	csReposURL.RawQuery = queryValues.Encode()
 
-	return csc.request("GET", csReposURL.String(), map[string]string{
+	resp, err := csc.request("GET", csReposURL.String(), map[string]string{
 		"x-rh-identity": id,
 	}, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode != http.StatusUnauthorized {
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("Unable to fetch repositories, got %v response, body: %s", resp.StatusCode, body)
+			}
+		}
+		return nil, fmt.Errorf("Unable to fetch repositories, got %v response", resp.StatusCode)
+	}
+
+	var repos *ApiRepositoryCollectionResponse
+	err = json.NewDecoder(resp.Body).Decode(&repos)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to parse repositories: %v", err)
+	}
+
+	return repos, nil
+}
+
+func (csc *ContentSourcesClient) GetRepositories(ctx context.Context, repoURLs []string, repoIDs []string, external bool) (RepositoryByID, error) {
+	result := RepositoryByID{}
```

Comment:
The map could be preallocated since the result capacity is already known, just a nitpick.